akka-cluster-sharding/src/main/scala/akka/cluster/sharding/internal/coordinator/CoordinatorRememberEntitiesStore.scala

```diff
@@ -0,0 +1,22 @@
+/*
+ * Copyright (C) 2009-2020 Lightbend Inc. <https://www.lightbend.com>
+ */
+package akka.cluster.sharding.internal.coordinator
+
+import akka.annotation.InternalApi
+import akka.cluster.sharding.ShardRegion.ShardId
+
+/**
+ * INTERNAL API
+ *
+ * Protocol for remember entities (shards) from the coordinator
+ */
+@InternalApi
+private[akka] object CoordinatorRememberEntitiesStore {
```

Comment:
Let's see, correct me if I'm wrong. The purpose of this store is to keep track of all shardIds that have ever been used.


When coordinator starts up it retrieves all ids from the store and starts the shards, so that they can start their entities. It's not critical for the initial coordinator state. It can be a background task to retrieve them.

When a previously unknown shardId is allocated (`ShardHomeAllocated`) it must add this to to the store before replying to the GetShardHome request.

Maybe we can skip the big refactoring of ShardCoordinator/PersistentShardCoordinator/DDataShardCoordinator now? Add the CoordinatorRememberEntitiesStore only and integrate those two interactions with the store into the existing code (including removing the old AllShards stuff from DDataShardCoordinator).

My thinking is that we don't want to keep PersistentShardCoordinator forever anyway, and the ShardCoordinator is not the area where we need to do changes for improving remember entities.

Reply:
I'm on the fence, I can see that it would be nice to avoid the risk of breaking something and avoiding extra work.
To arguments in favor of this separation though:

 * It'd be good to have the coordinator less complex for the parts where that is possible, given that it has some essential complexity we won't get away from
 * We'd still need to be able to implement the remember entities store _with the_ two current state stores if we keep those as concrete coordinator types, gut feeling is that would be much messier than having them separated

Reply:
> It'd be good to have the coordinator less complex for the parts where that is possible

Sure, but I think a better point of doing that is when we remove PersistentShardCoordinator. DDataShardCoordinator is implemented in similar event sourcing way, and that is probably not the way will do it when we only have one ShardCoordinator implementation. When we only have one implementation there is not much value of having the separation of the state store.

> We'd still need to be able to implement the remember entities store with the two current state stores

Can't PersistentShardCoordinator be left pretty much as is, and it implements the remember entities part already? We would add the integration of the new CoordinatorRememberEntitiesStore in the existing DDataShardCoordinator.

Reply:
Ok, as I said, on the fence, not strong feelings either way. I'll go with a more minimal impact something then.

Reply:
and I might be missing something (this was mostly a gut feeling comment)

Reply:
After working some on a smaller change in #28909 I think you are definitely right, good call. I'll close this and continue with that one instead.